lib/run_loop/app.rb

```diff
@@ -16,7 +16,21 @@ def initialize(app_bundle_path)
       @path = File.expand_path(app_bundle_path)
 
       if !App.valid?(app_bundle_path)
-        raise ArgumentError,
+        if App.cached_app_on_simulator?(app_bundle_path)
+          raise ArgumentError, %Q{
```

Comment:
ArgumentError seems like a slightly lazy choice of exception type? I would also assume that ArgumentError is a programming error from the callsite - that upfront validation can determine that the call is invalid; but I am not sure this is the case here..?
 I think a custom exception hierarchy could make sense perhaps? Maybe AppStateError as base class, and then raise a subclass ala "PartialAppInstallationError" here, and raise another subclass in the else-branch below; "AppNotFoundError" or somesuch.

Reply:
Will change to RuntimeError. 

When we encounter these kinds of problems, there is no retrying.

Reply:
> validation

All the validation is done in this method.  The caller does no validation.